src/AWS/Lambda/Combinators.hs

```diff
@@ -0,0 +1,185 @@
+{-|
+Module      : AWS.Lambda.Combinators
+Description : Function transformers that can be used to adapt the base runtime into other useful interfaces.
+Copyright   : (c) Nike, Inc., 2018
+License     : BSD3
+Maintainer  : nathan.fairhurst@nike.com, fernando.freire@nike.com
+Stability   : stable
+
+These combinators are for those who need to peek below the abstraction of the basic runtimes, for whatever reason.
+
+They map functions (instead of values) to turn basic handlers into handlers compatible with the base runtime.  These combinators allow us to expose functionality across many dimensions in an abstract way.  It also allows simple building blocks for those who need to "get in the middle" or adapt the basic runtimes in new ways without rebuilding everything from the ground up.
+-}
+
+module AWS.Lambda.Combinators (
+    withIOInterface,
+    withFallableInterface,
+    withPureInterface,
+    withoutContext
+) where
+
+import           Control.Monad.IO.Class (MonadIO, liftIO)
+import           Control.Monad.Reader   (MonadReader, ask)
+
+
+-- Helper for converting an either result into a monad/exception
+dropEither :: Monad m => Either String a -> m a
+dropEither = \case 
+     Left e  -> error e
```

Comment:
Is there a strong case to be made for `error` over the monadic [`fail`](http://hackage.haskell.org/package/base-4.12.0.0/docs/Prelude.html#v:fail)?

Reply:
I think there might be, but worth thinking about.  For the future of `fail` check this out:  https://prime.haskell.org/wiki/Libraries/Proposals/MonadFail

This will be a type constraint and not all Monads will actually have a `fail` instance.  Maybe that's not an issue?  And maybe our base handler should have a `MonadFail` instance instead of catching Exceptions?  Though, in reality, we must _inevitably_ catch exceptions at the top level, so it just makes it probably more complicated.  Hmm...

Reply:
Woah neat, thanks for linking to `MonadFail`, I wasn't aware of this! Seems like `error` is more in line with what we want right now then.

Not sure if it's worth breaking out the Either at this point, the composition doesn't look so ugly right now :P